### ISSUE TITLE
[BUG] Pomegranate 0.13.3 crashes when training a simple Bayes net

### ISSUE THREAD

**Issue by alexhenrie** [jmschrei/pomegranate#766]
It appears I spoke too soon when praising the virtues of version 0.13.3. After upgrading, I am no longer able to build my test Bayes net because Pomegranate crashes. Here's the code:

```
from neurtu import delayed, Benchmark
from pomegranate import BayesianNetwork
from sklearn.datasets import load_digits

X = load_digits(return_X_y=True)[0][:,:10]
net = BayesianNetwork().from_samples(X)
```

`git bisect` said:

```
b2cc6286b7d9fcf39b42c8b1742105b733c6e769 is the first bad commit
commit b2cc6286b7d9fcf39b42c8b1742105b733c6e769
Author: Jacob Schreiber <jmschreiber91@gmail.com>
Date:   Tue May 26 09:54:47 2020 -0700

    FIX log to log2 in BN

 pomegranate/BayesianNetwork.pyx | 18 +++++++++---------
 pomegranate/utils.pxd           |  1 +
 pomegranate/utils.pyx           |  9 ++++++++-
 3 files changed, 18 insertions(+), 10 deletions(-)
```

And indeed, the problem disappears after reverting that commit.

**Comment by jmschrei** [comment]
Can you show me what the error is? Things seemed fine on my end.

**Comment by jmschrei** [comment]
It looks like the problem is unrelated to the log but due to the consequences of having a large `max_parents`. Basically, if you set `max_parents=6`, which is what it used to be set to when doing the calculation with the natural log, the model will train. However, if you set `max_parents=7` or higher, the segfault is present. This occurs when using either the greedy algorithm or the A* exact algorithm. This suggested that there might be an issue with memory allocation. 

I then tried the dynamic programming exact algorithm, which is generally more resistant to bugs than the others (and, importantly, doesn't use the ParentGraph object), which returned the error 

```
  File "pomegranate/BayesianNetwork.pyx", line 2221, in pomegranate.BayesianNetwork.generate_parent_graph
    m[k+1] = m[k] * key_count_ndarray[variable]
OverflowError: value too large to convert to int
```

The error is occuring because you need the product of the number of keys per parent as the number of parameters for a variable (because you need to define the conditional probability table). The small increase in the number of parents from 6 to 7 caused an overflow issue. If you use a binarized version instead, i.e.

```python
X = (load_digits(return_X_y=True)[0][:,:10] > 2).astype('int32')
```

it completes immediately.

Basically, you've found an error, but it's related to an overflow somewhere else in the code that should be fixed, but doesn't need to be done immediately. 

Thanks for the report!

**Comment by alexhenrie** [comment]
Interesting. Thanks for the explanation!
